### PR TITLE
new pixycam protocol parser

### PR DESCRIPTION
**Pull request by jschall**

### CODE REVIEW

File: libraries/AP_IRLock/IRLock.h [hunk 1]
```diff
@@ -60,13 +61,14 @@ class IRLock
     uint32_t _last_update_ms;
 
     // irlock_target_info is a duplicate of the PX4Firmware irlock_s structure
-    typedef struct {
+    struct irlock_target_info {
```
Comment (jschall) [inline libraries/AP_IRLock/IRLock.h]: Add the typedef back in. This is declaring a variable irlock_target_info with type struct irlock_target_info { ... }. What you want to do is define the type irlock_target_info as struct { ... }.

So it should be:
```
typedef struct {
    // ...
} irlock_target_info;
irlock_target_info _target_info[10];
```

File: libraries/AP_IRLock/pixy_parser.cpp [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+    // printf("\n--------------------------AFTER EMPTY: ");
+    for (size_t i=0; i<pixy_len; i++) {
+        // printf("%u, ", (unsigned)pixy_buf[i]);
+    }
+}
+
+// Method - print_buffer
+void pixy_parser::print_buffer() {
+    struct blob_buffer& writebuf = blob_buffer[blob_buffer_write_idx];
+    // printf("Buffer:");
```
Comment (jschall) [inline libraries/AP_IRLock/pixy_parser.cpp:63]: Remove all commented code or place in #ifdefs

File: libraries/AP_IRLock/pixy_parser.h [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    pixy_buf[17]= {};
+    pixy_len = 0;
+    blob_buffer_write_idx = 0;
+    bytes_to_sof = 0;
+    bytes_to_block = 0;
+}
+*/
+class pixy_parser {
+    public:
+        bool if_swap_buffer;
```
Comment (jschall) [inline libraries/AP_IRLock/pixy_parser.h]: Avoid public member variables. I recommend making recv_byte_pixy() return true when the buffer is swapped.

File: libraries/AP_IRLock/pixy_parser.h [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+            MESSAGE_VALID_BLOCK
+        };
+
+        bool write_buffer(const pixy_blob& blob1);
+        enum message_validity_t check_pixy_message();
+
+        uint8_t pixy_buf[PIXY_PARSER_PIXY_BUF_SIZE];
+        size_t pixy_len;
+        uint8_t blob_buffer_write_idx;
+        size_t bytes_to_sof;
```
Comment (jschall) [inline libraries/AP_IRLock/pixy_parser.h]: If we're not using these two variables (IIRC we are not) we should just remove them.

File: libraries/AP_IRLock/pixy_parser.h [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+/*
+    pixy_buf_size = 17;
+    pixy_buf[17]= {};
+    pixy_len = 0;
+    blob_buffer_write_idx = 0;
+    bytes_to_sof = 0;
+    bytes_to_block = 0;
+}
+*/
```
Comment (jschall) [inline libraries/AP_IRLock/pixy_parser.h]: Remove unnecessary comment

File: libraries/AP_IRLock/AP_IRLock_I2C.cpp [hunk 6]
```diff
@@ -42,133 +39,119 @@ void AP_IRLock_I2C::init(int8_t bus)
...
-    struct frame irframe;
+
+                    // if (pixy parser just swapped buffers) {  ----------$$
+                    //    take semaphore                    -----------$$
+                    //    CONVERT frame
+                    // copy frame data from pixy parser to target info array
+                    //    update target_count
+                    //    update frame_timestamp_us    ->    uint32_t timestamp_us = micros();
+                    //    give semaphore
+        // }
```
Comment (jschall) [inline libraries/AP_IRLock/AP_IRLock_I2C.cpp]: Remove commented code.

File: libraries/AP_IRLock/AP_IRLock_I2C.cpp [hunk 7]
```diff
@@ -42,133 +39,119 @@ void AP_IRLock_I2C::init(int8_t bus)
...
-    float corner1_pos_x, corner1_pos_y, corner2_pos_x, corner2_pos_y;
-    pixel_to_1M_plane(corner1_pix_x, corner1_pix_y, corner1_pos_x, corner1_pos_y);
-    pixel_to_1M_plane(corner2_pix_x, corner2_pix_y, corner2_pos_x, corner2_pos_y);
+            const pixy_parser::pixy_blob* temp;
+            temp = pixyObj.read_buffer(0);
+
+            if (temp != nullptr) {
+                printf("\nBLOCK:- \nX: 0x%04x - Y: 0x%04x - W: 0x%04x - H: 0x%04x\n\n\n", temp->center_x, temp->center_y, temp->width, temp->height);
+
+//                convert_pixy_data(temp->center_x, temp->center_y, temp->width, temp->height);   //    CONVERT frame
```
Comment (jschall) [inline libraries/AP_IRLock/AP_IRLock_I2C.cpp]: Remove commented code.

File: libraries/AP_IRLock/AP_IRLock_I2C.cpp [hunk 8]
```diff
@@ -42,133 +39,119 @@ void AP_IRLock_I2C::init(int8_t bus)
...
+*/
+
+//}
+//    printf("\n&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&------------------OUT");
+//    printf("\nBLOCK:- \nX: 0x%04x - Y: 0x%04x - W: 0x%04x - H: 0x%04x\n\n\n", temp->center_x, temp->center_y, temp->width, temp->height);
+
+
+//
+//    }
+//
```
Comment (jschall) [inline libraries/AP_IRLock/AP_IRLock_I2C.cpp]: Remove commented code.

File: libraries/AP_IRLock/AP_IRLock_I2C.cpp [hunk 9]
```diff
@@ -42,133 +39,119 @@ void AP_IRLock_I2C::init(int8_t bus)
...
-bool AP_IRLock_I2C::update()
-{
+bool AP_IRLock_I2C::update() {
     bool new_data = false;
     if (!dev || !sem) {
         return false;
     }
     if (sem->take(HAL_SEMAPHORE_BLOCK_FOREVER)) {
-        if (_last_update_ms != _target_info.timestamp) {
+        if (_last_update_ms != _target_info[0].timestamp) {
```
Comment (jschall) [inline libraries/AP_IRLock/AP_IRLock_I2C.cpp]: Don't duplicate timestamp. Have a single _frame_timestamp variable.

File: libraries/AP_IRLock/AP_IRLock_I2C.h [hunk 10]
```diff
@@ -15,26 +16,16 @@ class AP_IRLock_I2C : public IRLock
...
-    };
-
-    bool timer(void);
-
-    bool sync_frame_start(void);
-    bool read_block(struct frame &irframe);
     void read_frames(void);
-
     void pixel_to_1M_plane(float pix_x, float pix_y, float &ret_x, float &ret_y);
+//    void convert_pixy_data(uint16_t x, uint16_t y, uint16_t w, uint16_t h);
```
Comment (jschall) [inline libraries/AP_IRLock/AP_IRLock_I2C.h]: Remove commented code.

File: libraries/AP_IRLock/pixy_parser.h (hunk 5 above)
Comment (Aksellence) [reply libraries/AP_IRLock/pixy_parser.h]: Done

File: libraries/AP_IRLock/AP_IRLock_I2C.cpp (hunk 6 above)
Comment (Aksellence) [reply libraries/AP_IRLock/AP_IRLock_I2C.cpp]: Done

File: libraries/AP_IRLock/AP_IRLock_I2C.h (hunk 10 above)
Comment (Aksellence) [reply libraries/AP_IRLock/AP_IRLock_I2C.h]: Done

File: libraries/AP_IRLock/IRLock.h (hunk 1 above)
Comment (Aksellence) [reply libraries/AP_IRLock/IRLock.h]: Done

File: libraries/AP_IRLock/AP_IRLock_I2C.cpp (hunk 7 above)
Comment (Aksellence) [reply libraries/AP_IRLock/AP_IRLock_I2C.cpp]: Done

File: libraries/AP_IRLock/AP_IRLock_I2C.cpp (hunk 8 above)
Comment (Aksellence) [reply libraries/AP_IRLock/AP_IRLock_I2C.cpp]: Done

File: libraries/AP_IRLock/AP_IRLock_I2C.cpp (hunk 9 above)
Comment (Aksellence) [reply libraries/AP_IRLock/AP_IRLock_I2C.cpp]: Done

File: libraries/AP_IRLock/pixy_parser.h (hunk 4 above)
Comment (Aksellence) [reply libraries/AP_IRLock/pixy_parser.h]: Done

File: libraries/AP_IRLock/pixy_parser.h (hunk 3 above)
Comment (Aksellence) [reply libraries/AP_IRLock/pixy_parser.h]: Done

File: libraries/AP_IRLock/pixy_parser.cpp (hunk 2 above)
Comment (Aksellence) [reply libraries/AP_IRLock/pixy_parser.cpp:63]: Done. Left print statements for testing

**Comment by Aksellence** [comment]
Made required changes

**Pull request closed without merging**